include/unifex/linux/io_epoll_context.hpp

```diff
@@ -477,6 +499,400 @@ inline io_epoll_context::scheduler io_epoll_context::get_scheduler() noexcept {
+      if (result == -ECANCELED) {
+        unifex::set_done(std::move(self.receiver_));
+      } else if (result >= 0) {
+        unifex::set_value(std::move(self.receiver_), ssize_t(result));
+      } else {
+        printf("on_read_complete::readv failed\n");
+        unifex::set_error(
+            std::move(self.receiver_),
+            std::error_code{-int(result), std::system_category()});
+      }
+    }
+
+    static void complete_with_done(operation_base* op) noexcept {
+      auto& self = *static_cast<operation*>(op);
+      self.execute_ = nullptr;
+      self.request_stop_local();
+    }
+
+    void request_stop() noexcept {
+      if (context_.is_running_on_io_thread()) {
+        request_stop_local();
+      } else {
+        request_stop_remote();
+      }
+    }
+
+    void request_stop_local() noexcept {
+      assert(context_.is_running_on_io_thread());
+
+      stopCallback_.destruct();
+
+      epoll_event event = {};
+      (void)epoll_ctl(context_.epollFd_.get(), EPOLL_CTL_DEL, fd_, &event);
+      execute_ = nullptr;
+
+      // Avoid instantiating set_done() if we're not going to call it.
+      if constexpr (is_stop_ever_possible) {
+        unifex::set_done(std::move(receiver_));
+      } else {
+        // This should never be called if stop is not possible.
+        assert(false);
+      }
+    }
+
+    void request_stop_remote() noexcept {
+      epoll_event event = {};
+      (void)epoll_ctl(context_.epollFd_.get(), EPOLL_CTL_DEL, fd_, &event);
+
+      this->execute_ = &operation::complete_with_done;
+      this->context_.schedule_remote(this);
+    }
+
+    struct cancel_callback {
+      operation& op_;
+
+      void operator()() noexcept {
+        op_.request_stop();
+      }
+    };
+
+    io_epoll_context& context_;
+    int fd_;
+    iovec buffer_[1];
+    Receiver receiver_;
+    manual_lifetime<typename stop_token_type_t<
+      Receiver>::template callback_type<cancel_callback>>
+      stopCallback_;
+  };
+
+ public:
+  // Produces number of bytes read.
+  template <
+      template <typename...> class Variant,
+      template <typename...> class Tuple>
+  using value_types = Variant<Tuple<ssize_t>>;
+
+  template <template <typename...> class Variant>
+  using error_types = Variant<std::error_code>;
+
+  explicit read_sender(
+      io_epoll_context& context,
+      int fd,
+      span<std::byte> buffer) noexcept
+      : context_(context), fd_(fd), buffer_(buffer) {}
+
+  template <typename Receiver>
+  operation<std::decay_t<Receiver>> connect(Receiver&& r) && {
+    return operation<std::decay_t<Receiver>>{*this, (Receiver &&) r};
+  }
+  template <typename Receiver>
+  operation<std::decay_t<Receiver>> connect(Receiver&& r) & {
+    return operation<std::decay_t<Receiver>>{*this, (Receiver &&) r};
+  }
+  template <typename Receiver>
+  operation<std::decay_t<Receiver>> connect(Receiver&& r) const & {
+    return operation<std::decay_t<Receiver>>{*this, (Receiver &&) r};
+  }
```

Comment:
ok. this increases my confidence that this form of read and write is not the correct basis.
we read sequences of buffers, not one buffer. having to add defer inside of each repeat is a lot of noise.

Reply:
I'd remind you that i/o is done in terms of `struct iovec[N]` in any case, sequences of buffers are a (poor) abstraction over that, because they don't sufficiently encourage arrays of `struct iovec`. Also, reads from files are **not** idempotent, file content and length can always asynchronously change at any moment for any reason. They are thus never repeatable, ever, and to ever treat them as such leads to loss of correctness. The only atomicity and concurrency guarantees you get are those given by POSIX on a per `read()` and `write()` operation, up to per-syscall the-system-at-the-time limits.